File: GoLang/small_codes/error_handling.go
package main
import (
    "errors"
    "fmt"
)

func divide(a, b float64) (float64, error) {
    if b == 0 {
        return 0, errors.New("division by zero")
    }
    return a / b, nil
}

func main() {
    res, err := divide(10, 5)
    if err != nil {
        fmt.Println("Error:", err)
    } else {
        fmt.Println("Result:", res)
    }

    res1, err1 := divide(10, 0)
    if err1 != nil {
        fmt.Println("Error:", err1)
    } else {
        fmt.Println("Result:", res1)
    }
}

File: Rust/small_codes/error_handling.rs
// Enhanced Error Handling Demo
fn divide(a: f64, b: f64) -> Result<f64, String> {
    if b == 0.0 {
        Err("Division by zero is not allowed".into())
    } else {
        Ok(a / b)
    }
}

fn main() {
    // Demonstrate Result handling
    let divisions = vec![(10.0, 2.0), (5.0, 0.0), (7.0, 3.0)];
    for (a, b) in divisions {
        match divide(a, b) {
            Ok(result) => println!("{} ÷ {} = {:.2}", a, b, result),
            Err(err) => println!("Failed to divide {} by {}: {}", a, b, err),
        }
    }

    // Demonstrate Option with unwrap_or
    let some_value: Option<i32> = Some(42);
    let none_value: Option<i32> = None;

    println!("Some value: {}", some_value.unwrap_or(0));
    println!("None value fallback: {}", none_value.unwrap_or(-1));

    // Using unwrap_or_else for dynamic fallback
    let dynamic_fallback = none_value.unwrap_or_else(|| {
        println!("Computing fallback value...");
        100
    });
    println!("Dynamic fallback: {}", dynamic_fallback);
}